### PR TITLE
RM-7050: Add Firefox support

### PR DESCRIPTION
**Pull request by iam-mholle**
https://re-motion.atlassian.net/browse/RM-7050

### CODE REVIEW

File: Remotion/ObjectBinding/Web.Development.WebTesting.IntegrationTests/BocEnumValueControlObjectTest.cs [hunk 1]
```diff
@@ -461,6 +461,9 @@ public void TestSelectOption ()
     [Test]
     public void TestSelectOptionPostBack ()
     {
+      if (Helper.BrowserConfiguration.IsFirefox())
```
Comment (MichaelKetting) [inline Remotion/ObjectBinding/Web.Development.WebTesting.IntegrationTests/BocEnumValueControlObjectTest.cs]: when will this get fixed? What are we doing with this issue when writing webtests?

File: Remotion/ObjectBinding/Web.Development.WebTesting.IntegrationTests/ObjectBinding.Web.Development.WebTesting.IntegrationTests.csproj [hunk 2]
```diff
@@ -178,15 +178,19 @@
...
-  </ItemGroup>
-  <ItemGroup>
     <EmbeddedResource Include="SavedTestScreenshots\BocListControlObjectTest\InternetExplorer.any.ScreenshotTest_TableContainer3.png" />
-  </ItemGroup>
-  <ItemGroup>
     <EmbeddedResource Include="SavedTestScreenshots\BocTreeViewControlObjectTest\Chrome.any.WebTreeView.png" />
-  </ItemGroup>
-  <ItemGroup>
     <EmbeddedResource Include="SavedTestScreenshots\BocAutoCompleteReferenceValueControlObjectTest\InternetExplorer.any.ScreenshotTest2.png" />
+    <EmbeddedResource Include="SavedTestScreenshots\BocAutoCompleteReferenceValueControlObjectTest\Firefox.any.ScreenshotTest.png" />
```
Comment (MichaelKetting) [inline Remotion/ObjectBinding/Web.Development.WebTesting.IntegrationTests/ObjectBinding.Web.Development.WebTesting.IntegrationTests.csproj]: quick demo of the firefox screenshots please. possibly compared to the chrome ones.

File: Remotion/ObjectBinding/Web.Development.WebTesting/ControlObjects/BocAutoCompleteReferenceValueControlObject.cs [hunk 3]
```diff
@@ -352,7 +352,8 @@ private SearchServiceResultItem GetFirstAutoCompleteResult ([NotNull] string fil
   if(returnValue.state === 'success' && !Array.isArray(returnValue.data))
     returnValue.data = [ returnValue.data ];
 
-  return returnValue;
+  // Strip methods from return object to prevent unwanted errors from Firefox Marionette.
+  return JSON.parse(JSON.stringify(returnValue));
```
Comment (MichaelKetting) [inline Remotion/ObjectBinding/Web.Development.WebTesting/ControlObjects/BocAutoCompleteReferenceValueControlObject.cs]: What the...this is sooo weird, this stuff.

File: Remotion/Web/Development.WebTesting.IntegrationTests/BrowserSessionTest.cs [hunk 4]
```diff
@@ -70,6 +70,22 @@ public void ChromeDriver_SupportsBrowserLogs ()
       Assert.That (logs.Count (l => l.Message.Contains (warningMessage)), Is.EqualTo (1));
     }
 
+    [Test]
+    public void GeckoDriver_DoesNotSupportBrowserLogs ()
+    {
+      if (!Helper.BrowserConfiguration.IsFirefox())
```
Comment (MichaelKetting) [inline Remotion/Web/Development.WebTesting.IntegrationTests/BrowserSessionTest.cs]: Okay, right now this only affects the trunk/devlop. But we will need some logic here to deal with browser logs in firefox.

File: Remotion/Web/Development.WebTesting.IntegrationTests/MultiWindowTest.cs [hunk 5]
```diff
@@ -123,6 +127,9 @@ public void TestMultiWindowActions ()
     [Test]
     public void TestAcceptModalBrowserDialog ()
     {
+      if (Helper.BrowserConfiguration.IsFirefox())
```
Comment (MichaelKetting) [inline Remotion/Web/Development.WebTesting.IntegrationTests/MultiWindowTest.cs]: I thought we only had 1 issue?

File: Remotion/Web/Development.WebTesting/BrowserSession/BrowserSessionBase.cs [hunk 6]
```diff
@@ -120,7 +120,10 @@ public virtual void Dispose ()
 
       // Check driver and main browser for null
       if (driverProcess != null)
+      {
         ProcessUtils.GracefulProcessShutdown (driverProcess, c_driverShutdownWaitTime);
+        driverProcess.WaitForExit();
```
Comment (MichaelKetting) [inline Remotion/Web/Development.WebTesting/BrowserSession/BrowserSessionBase.cs]: we probbly need a timeout here. 1 minute is sufficient?

File: Remotion/Web/Development.WebTesting.IntegrationTests/BrowserSessionTest.cs (hunk 4 above)
Comment (MichaelKetting) [reply Remotion/Web/Development.WebTesting.IntegrationTests/BrowserSessionTest.cs]: Okay, seen production code. please change to use two paths in the test, one for firefox, one for the rest and asser tthe getlog() result in both cases.

File: Remotion/Web/Development.WebTesting/WebDriver/Configuration/Firefox/FirefoxBinariesProvider.cs [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+          string.Format (
+              "Could not fetch the driver's download URL from GitHub. This could mean GitHub's API has changed, or the release does not contain the win{0} archive.",
+              c_driverBitness));
+    }
+
+    private GithubResponse GetLatestGeckoDriverReleaseInfo ()
+    {
+      using (var webClient = new WebClient())
+      {
+        webClient.Headers.Add ("user-agent", "re-motion");
```
Comment (MichaelKetting) [inline Remotion/Web/Development.WebTesting/WebDriver/Configuration/Firefox/FirefoxBinariesProvider.cs]: we really need a user agent? Is that also there with the other providers? can we remove that in all providers (not in this branch)?

Comment (iam-mholle) [reply Remotion/Web/Development.WebTesting/WebDriver/Configuration/Firefox/FirefoxBinariesProvider.cs]: The user agent is required for github requests. Other providers do not need the agent set as they do not issue github requests.

**Comment by iam-mholle** [comment]
Added to master with 29e2af7260fc9748630c7bc92c84cfe3a8586a67.

**Pull request closed without merging**